mono/metadata/marshal.c

```diff
@@ -1743,14 +1743,26 @@ emit_object_to_ptr_conv (MonoMethodBuilder *mb, MonoType *type, MonoMarshalConv
 	}
 }
 
+static int offset_of_first_nonstatic_field(MonoClass *klass)
+{
+	int i;
+	for (i = 0; i < klass->field.count; i++)
+	{
+		if (!(klass->fields[i].type->attrs & FIELD_ATTRIBUTE_STATIC) && !mono_field_is_deleted(&klass->fields[i]))
+			return klass->fields[i].offset - sizeof(MonoObject);
```

Comment:
Because the offsets are added to the MonoObject header when settings up the class fields with SEQUENTIAL_LAYOUT.

https://github.com/Unity-Technologies/mono/blob/unity-staging/mono/metadata/class.c#L1677
https://github.com/Unity-Technologies/mono/blob/unity-staging/mono/metadata/class.c#L1721
https://github.com/Unity-Technologies/mono/blob/unity-staging/mono/metadata/class.c#L1728